[Dashboards] [Embeddables] Allow embeddables to register a panel config schema

nickpeihl opened the pull request:
This is a WIP example for allowing embeddables to register a panel config schema that allows the Dashboard to validate panel content via the public HTTP or internal RPC CRUD endpoints.


Initially, this adds validation for Lens data table charts. An example of a chart that gets validated against the schema is below. Other Lens charts will be validated against a generic schema. Embeddables that do not register a schema are not validated.

```
curl --request POST \
  --url http://localhost:5601/api/dashboards/dashboard \
  --header 'accept: application/json' \
  --header 'authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'elastic-api-version: 2023-10-31' \
  --header 'kbn-xsrf: true' \
  --data '{"attributes": {"title": "a table lens panel","panels": [{"type": "lens","gridData": {"x": 0,"y": 0,"w": 24,"h": 15},"panelIndex": "666ba2ac-babb-4226-ad8b-9e30fec7dd74","panelConfig": {"attributes": {"visualizationType": "lnsDatatable","title": "my data table","state": {"datasourceStates": {"formBased": {"layers": {"53c6e98c-6267-4d97-876b-29f5df00e39b": {"columns": {"963714a8-b8a2-4e7d-889e-3851c74a0212": {"dataType": "string","isBucketed": true,"operationType": "terms","sourceField": "machine.os.keyword","params": {"size": 5,"orderBy": {"type": "alphabetical"}}},"627d787d-f907-46d2-861f-7752cd1001b9": {"dataType": "number","isBucketed": false,"label": "Count of records","operationType": "count","sourceField": "___records___"}},"columnOrder": ["963714a8-b8a2-4e7d-889e-3851c74a0212","627d787d-f907-46d2-861f-7752cd1001b9"]}}}},"visualization": {"layerId": "53c6e98c-6267-4d97-876b-29f5df00e39b","layerType": "data","columns": [{"columnId": "627d787d-f907-46d2-861f-7752cd1001b9"},{"columnId": "963714a8-b8a2-4e7d-889e-3851c74a0212"}]},"query": {"query": "","language": "kuery"},"filters": []},"references": [{"type": "index-pattern","id": "90943e30-9a47-11e8-b64d-95841ca0b247","name": "indexpattern-datasource-layer-53c6e98c-6267-4d97-876b-29f5df00e39b"}]}}}]},"references": [{"type": "index-pattern","id": "90943e30-9a47-11e8-b64d-95841ca0b247","name": "666ba2ac-babb-4226-ad8b-9e30fec7dd74:indexpattern-datasource-layer-53c6e98c-6267-4d97-876b-29f5df00e39b"}]}'
```


## Summary

Summarize your PR. If it involves visual changes include a screenshot or gif.


### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

### Identify risks

Does this PR introduce any risks? For example, consider risks like hard to test bugs, performance regression, potential of data loss.

Describe the risk, its severity, and mitigation for each identified risk. Invite stakeholders and evaluate how to proceed before merging.

- [ ] [See some risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx)
- [ ] ...

nickpeihl committed:
Allow embeddables to register a panel config schema

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

nickpeihl committed:
First attempt at UnionType for panelConfig

This fails during validation due to mis-matched properties between embeddable schemas.

nickpeihl committed:
Embeddable schemas as part of Open API spec

nickpeihl commented:
Commit a30830b shows an example of how embeddables can register a schema that we can use when validating request and response bodies. Specifically, the validation is performed by adding a [custom `validate` function](https://github.com/elastic/kibana/pull/201474/commits/a30830b651034c137010cd8862d99707b5f46fda?diff=unified&w=0#diff-692f4b808f9f013cc21265b394b035af3a666f054d941817c09cd1005563abe2R322-R333) to the panel schema. While this works for validation, these schemas are not exposed to the API routes so the Open API generator does not receive the embeddable schemas.

A few attempts were made to provide the embeddable schemas to the routes and the generated Open API specification. 

## Attempt 1 - Create a UnionType of schemas registered on the embeddable plugin
This would have been the simplest as we could pass an array of the registered schemas to a `schema.oneOf`. However, this failed presumably due to the differences in properties between different embeddable schemas.


<details>
<summary>Here's a unit test showing the behavior.</summary>

```typescript
test('UnionType that validates against the links schema and a couple of lens chart types', () => {
  const genericPanelConfig = schema.object({
    version: schema.maybe(
      schema.string({
        meta: { description: 'The version of the embeddable in the panel.' },
      })
    ),
    title: schema.maybe(schema.string({ meta: { description: 'The title of the panel' } })),
    description: schema.maybe(
      schema.string({ meta: { description: 'The description of the panel' } })
    ),
    savedObjectId: schema.maybe(
      schema.string({
        meta: {
          description: 'The unique id of the library item to construct the embeddable.',
        },
      })
    ),
    hidePanelTitles: schema.maybe(
      schema.boolean({
        defaultValue: false,
        meta: { description: 'Set to true to hide the panel title in its container.' },
      })
    ),
    enhancements: schema.maybe(schema.recordOf(schema.string(), schema.any())),
    disabledActions: schema.maybe(schema.arrayOf(schema.string())),
    disableTriggers: schema.maybe(schema.boolean()),
    timeRange: schema.maybe(schema.object({ from: schema.string(), to: schema.string() })),
  });
  const linksPanelConfig = genericPanelConfig.extends({
    attributes: schema.object({
      layout: schema.maybe(
        schema.oneOf([schema.literal('horizontal'), schema.literal('vertical')])
      ),
      links: schema.arrayOf(
        schema.object({
          type: schema.oneOf([schema.literal('dashboardLink'), schema.literal('externalLink')]),
          id: schema.string(),
          order: schema.number(),
          destinationRefName: schema.maybe(schema.string()),
          destination: schema.maybe(schema.string()),
        })
      ),
    }),
  });
  const panelConfig = schema.oneOf([
    genericPanelConfig.extends({
      attributes: schema.object({
        visualizationType: schema.literal('lnsXY'),
        state: schema.recordOf(schema.string(), schema.any()),
      }),
    }),
    genericPanelConfig.extends({
      attributes: schema.object({
        visualizationType: schema.literal('lnsMetric'),
        state: schema.recordOf(schema.string(), schema.any()),
      }),
    }),
    linksPanelConfig,
  ]);

  expect(() =>
    panelConfig.validate({
      attributes: {
        visualizationType: 'lnsXY',
        state: {},
      },
    })
  ).not.toThrow();

  // This should throw because `lsnDatatable` is not a valid visualizationType on the schema
  expect(() =>
    panelConfig.validate({ attributes: { visualizationType: 'lnsDatatable', state: {} } })
  ).toThrow();

  // We would expect this to validate, but it throws an error.
  // "Cannot destructure property 'Symbol(Symbol.iterator)' of 'undefined' as it is undefined."
  // This seems to happen because the Lens charts schemas have a LiteralType `visualizationType` that
  // does not exist in the Links schema. 
  expect(() =>
    panelConfig.validate({
      attributes: {
        layout: 'horizontal',
        links: [
          {
            type: 'dashboardLink',
            id: 'test',
            order: 1,
            destinationRefName: 'test_ref',
          },
        ],
      },
    })
  ).not.toThrow();
});
```
</details>


## Attempt 2 - Using nested ConditionalType

This is more of an attempt to create a discriminated union using the kbn-config-schema where the `panels[0].type` is checked against the registered embeddable type. If it matches, the associated schema is used for validation. If it does not match, then it tries another conditional type.

This was implemented in [17ce0f8](https://github.com/elastic/kibana/pull/201474/commits/17ce0f85cee0621ff7338e80a96cc8c19d3d1af3).

<details>
<summary>Here is a unit test showing the ConditionalType behavior</summary>

```typescript
function buildConditionalSchema(conditions: Array<{ type: string; schema: any }>) {
  return conditions.reduce((acc, condition) => {
    return schema.conditional(
      schema.siblingRef('.type'), // Retrieves the value of the `type` property from the sibling fields
      condition.type, // The value to compare against the retrieved `type` property
      condition.schema, // The schema to use if the `type` property matches `condition.type`
      acc // A nested conditional schema to use if the `type` property does not match `condition.type`
    );
  }, schema.never()); // Initial schema to use if no conditions match
}

test('discriminated union', () => {
  enum ViewMode {
    EDIT = 'edit',
    PREVIEW = 'preview',
    PRINT = 'print',
    VIEW = 'view',
  }

  const gridDataSchema = schema.object({
    x: schema.number({ meta: { description: 'The x coordinate of the panel in grid units' } }),
    y: schema.number({ meta: { description: 'The y coordinate of the panel in grid units' } }),
    w: schema.number({
      defaultValue: 24,
      min: 1,
      max: 48,
      meta: { description: 'The width of the panel in grid units' },
    }),
    h: schema.number({
      defaultValue: 15,
      min: 1,
      meta: { description: 'The height of the panel in grid units' },
    }),
    i: schema.string({
      meta: { description: 'The unique identifier of the panel' },
      defaultValue: uuidv4(),
    }),
  });

  const commonPanelSchema = schema.object({
    gridData: gridDataSchema,
    panelIndex: schema.string({
      meta: { description: 'The unique ID of the panel.' },
      defaultValue: schema.siblingRef('gridData.i'),
    }),
    panelRefName: schema.maybe(schema.string()),
    id: schema.maybe(schema.string()),
    version: schema.maybe(schema.string()),
    title: schema.maybe(schema.string()),
  });

  const commonPanelConfigSchema = schema.object({
    version: schema.maybe(
      schema.string({
        meta: { description: 'The version of the embeddable in the panel.' },
      })
    ),
    viewMode: schema.maybe(
      schema.oneOf(
        Object.values(ViewMode)
          .filter((value): value is ViewMode => typeof value === 'string')
          .map((value) => schema.literal(value)) as [Type<ViewMode>]
      )
    ),
    title: schema.maybe(schema.string({ meta: { description: 'The title of the panel' } })),
    description: schema.maybe(
      schema.string({ meta: { description: 'The description of the panel' } })
    ),
    savedObjectId: schema.maybe(
      schema.string({
        meta: {
          description: 'The unique id of the library item to construct the embeddable.',
        },
      })
    ),
    hidePanelTitles: schema.maybe(
      schema.boolean({
        defaultValue: false,
        meta: { description: 'Set to true to hide the panel title in its container.' },
      })
    ),
    enhancements: schema.maybe(schema.recordOf(schema.string(), schema.any())),
    disabledActions: schema.maybe(schema.arrayOf(schema.string())),
    disableTriggers: schema.maybe(schema.boolean()),
    timeRange: schema.maybe(schema.object({ from: schema.string(), to: schema.string() })),
  });

  const conditions = [
    {
      type: 'lens',
      schema: commonPanelSchema.extends({
        type: schema.literal('lens'),
        panelConfig: commonPanelConfigSchema.extends({
          attributes: schema.object({
            visualizationType: schema.oneOf([schema.literal('lnsXY'), schema.literal('lnsMetric')]),
            state: schema.conditional(
              schema.siblingRef('visualizationType'),
              'lnsXY',
              schema.recordOf(schema.string(), schema.any()),
              schema.conditional(
                schema.siblingRef('visualizationType'),
                'lnsMetric',
                schema.recordOf(schema.string(), schema.any()),
                schema.never()
              )
            ),
          }),
        }),
      }),
    },
    {
      type: 'links',
      schema: commonPanelSchema.extends({
        type: schema.literal('links'),
        panelConfig: commonPanelConfigSchema.extends({
          attributes: schema.object({
            layout: schema.maybe(
              schema.oneOf([schema.literal('horizontal'), schema.literal('vertical')])
            ),
            links: schema.arrayOf(
              schema.object({
                type: schema.oneOf([
                  schema.literal('dashboardLink'),
                  schema.literal('externalLink'),
                ]),
                id: schema.string(),
                order: schema.number(),
                destinationRefName: schema.maybe(schema.string()),
                destination: schema.maybe(schema.string()),
              })
            ),
          }),
        }),
      }),
    },
  ];

  const type = buildConditionalSchema(conditions);

  const commonPanel = {
    gridData: { x: 1, y: 1, w: 1, h: 1, i: '1' },
    panelIndex: '1',
  };

  expect(() =>
    type.validate({
      type: 'lens',
      ...commonPanel,
      panelConfig: { attributes: { visualizationType: 'lnsXY', state: {} } },
    })
  ).not.toThrow();

  expect(() =>
    type.validate({
      type: 'lens',
      ...commonPanel,
      panelConfig: { attributes: { visualizationType: 'lnsMetric', state: {} } },
    })
  ).not.toThrow();

  // This should throw because `lnsSankey` is not a registered visualization type
  expect(() =>
    type.validate({
      type: 'lens',
      ...commonPanel,
      panelConfig: { attributes: { visualizationType: 'lnsSankey', state: {} } },
    })
  ).toThrow();

  expect(() =>
    type.validate({
      type: 'links',
      ...commonPanel,
      panelConfig: {
        attributes: {
          layout: 'horizontal',
          links: [
            {
              type: 'dashboardLink',
              id: '1',
              order: 1,
            },
          ],
        },
      },
    })
  ).not.toThrow();
});
```

</details>
